contracts/Removal.sol

```diff
@@ -316,6 +335,85 @@ contract Removal is
     });
   }
 
+  /**
+   * @notice Transfers the provided `amounts` of the specified removal `ids` directly to the Certificate contract to
+   * mint a legacy certificate.
+   *
+   * @dev The Certificate contract implements `onERC1155BatchReceived`, which is invoked upon receipt of any removals.
+   * This function circumvents the market contract's lifecycle. This function first transfers the removals from the
+   * supplier accounts to a single account that has the consignor role-- this is necessary because
+   * `_safeBatchTransferFrom` only accepts one `from` address and `Certificate.onERC1155BatchReceived` will mint a *new*
+   * certificate every time an additional batch is received.
+   *
+   * ##### Requirements:
+   * - Can only be used when the caller has the `CONSIGNOR_ROLE`
+   * - Can only be used when this contract is not paused
+   * - IDs must already have been minted via `mintBatch`.
+   * - Enforces the rules of `Removal._beforeTokenTransfer`.
+   *
+   * @param owners The owners of the specified removal IDs to transfer from.
+   * @param ids 2D array of the removal IDs to add to transfer to the Certificate contract. Each row in the array
+   * represents a list of one owners removal IDs.
+   * @param amounts 2D array of the balances of each token ID to transfer to the Certificate contract. Each row in the
+   * array represents a list of one owner's removal amounts, each of which corresponds to the same row and column
+   * defined in `ids`.
+   * @param certificateRecipient The recipient of the certificate to be minted.
+   * @param certificateAmount TThe total amount of the certificate.
+   */
+  function migrate(
```

Comment:
Given that the owner address is encoded in the recipient id aren't `owners` and the the first dimension of `ids` redundant ?

Reply:
It certainly makes the code short given this structure but we'd need to validate the arguments.

Reply:
it would require yet another loop where we have to loop through all the ids and sort them on-chain to get rebuild this argument. Is it really worth it given this function is already extemely expensive (worried about gas limits, not cost) as is?

Reply:
as it stands, it's about 2.5m gas for a call to `migrate` that spans 5 removals across 2 suppliers.

Reply:
Well, let's test with the worst case (well known) certificate and make decisions on the need for optimization based on that.

It's not pretty as it but if it does that trick and only gets used once nbd.

Reply:
Just ran a test based on this data you sent over: 
```json
{
  "certificateId": "5962695850328064",
  "certificateUri": "https://nori.com/certificate/1394",
  "removalCount": "190",
  "NRTs": "15000.0",
  "projecCount": "7"
}
```
and we're sitting at 44m gas (~double the block gas limit). Even as is we need to optimize for this to work.